aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSFileIO.java

```diff
@@ -111,7 +111,8 @@ public void initialize(Map<String, String> properties) {
       this.metrics = ctor.newInstance("oss");
 
       metrics.initialize(properties);
-    } catch (NoSuchMethodException | ClassCastException e) {
+    } catch (NoClassDefFoundError | NoSuchMethodException | ClassCastException e) {
+      this.metrics = MetricsContext.nullMetrics();
```

Comment:
Isn't this already set? I don't think we need this.

Reply:
Actually that is being overridden in the current logic. I have added some in-line comments below to highlight the issue:

```
private MetricsContext metrics = MetricsContext.nullMetrics(); // metrics initialised as nullMetrics at class level
...
...
  @Override
  public void initialize(Map<String, String> properties) {
...
...
    try {
      DynConstructors.Ctor<MetricsContext> ctor =
          DynConstructors.builder(MetricsContext.class).hiddenImpl(DEFAULT_METRICS_IMPL, String.class).buildChecked();
      this.metrics = ctor.newInstance("ecs"); // metrics re-initialised as new HadoopMetricsContext("ecs")

      metrics.initialize(properties); // throws exception in case of classpath issues 
    } catch (NoClassDefFoundError | NoSuchMethodException | ClassCastException e) {
      // We swallow the exception and allow new HadoopMetricsContext() to flow down to EcsInputFile/EcsOutputFile
      LOG.warn("Unable to load metrics class: '{}', falling back to null metrics", DEFAULT_METRICS_IMPL, e);
    }
  }
```

We can handle the fallback in the catch block (do suggest if there is some better way). Without that the CI seems to fail for commit: [c7e355f](https://github.com/apache/iceberg/pull/4489/commits/c7e355fc0820c5630218ac0e5ae1f125852e210a)

The CI passes if we handle the nullMetrics fallback in the catch block for commit: [2ad9cda](https://github.com/apache/iceberg/pull/4489/commits/2ad9cdae82ed3714ee58f96abfc95daf1b880a58)

Reply:
Do suggest if there is some better way to handle the fallback. I can implement the same for all the FileIO's in this PR.

Reply:
What about waiting until after `initialize` succeeds to set `this.metrics`? That seems better to me.

Reply:
If I understand this correctly:

```
private MetricsContext metrics = MetricsContext.nullMetrics();
...
try {
....
metrics.initialize(properties);
this.metrics = ctor.newInstance("scheme");
} catch() {
...
}
```

Reply:
No, you need to initialize a `MetricsContext` and then set `this.metrics` to the _initialized_ context.

If it were me, I would use add a local variable. That can keep track of the context and then you can update the instance field afterward:

```java
MetricsContext context = ctor.newInstance("scheme");
context.initialize(properties);
this.metrics = context;
```

The local variable helps you avoid modifying `metrics` until after `initialize` is called, but you still use the initialized metrics context instead of a completely new and uninitialized one.

Reply:
Ah, thanks for the detailed response. Yes, using the local variable makes perfect sense!